### PR TITLE
[jireh0108] iP

### PR DESCRIPTION
**Pull request by jireh0108**
### **Nova**

> "Productivity is never an accident. It’s always the result of a commitment to excellence, intelligent planning, and focused effort." – Paul J. Meyer ([source](https://instepuk.com/article/time-management/))

Nova frees your mind from trying to remember everything. It’s:

- Text-based – simple, clean, distraction-free
- Intuitive – easy to learn and use
- Lightning fast ⚡ – manage your tasks in seconds

**Getting started is as easy as 1-2-3:**

1. Download Nova from [here](https://github.com/jireh0108/ip).
2. Open it.
3. Add your tasks.
4. Let Nova handle the rest 😉

And yes—it’s completely **FREE**!

**Why you’ll love Nova:**

- [x] Keep your tasks organized effortlessly
- [ ] Track deadlines and never miss a due date (coming soon!)
- [ ] Set reminders to stay on top of your schedule (coming soon!)

For Java programmers: Here’s the `main` method to launch Nova:

```
public class Main {
    public static void main(String[] args) {
        new Nova("data/tasks.txt").run();
    }
}
```

**Coming soon:**

- Cross-device sync so you can access your tasks anywhere
- Smart suggestions to help you prioritize tasks
- Customizable themes for a personal touch

Nova isn’t just a task manager—it’s your productivity companion, ready to help you focus on what matters most.

### CODE REVIEW

File: src/main/java/nova/exceptions/NovaException.java
Comment (grenn24) [inline src/main/java/nova/exceptions/NovaException.java:1]: I think NovaException should extend Exception rather than RuntimeException, since most errors in a chatbot (like invalid user input or failed commands) are recoverable and should be explicitly handled by the caller. Currently, it only stores a message, but it could be more useful if it included additional fields such as the error code or the recovery process, which would make debugging and the error print statement more informative.

File: src/main/java/nova/Nova.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    private Ui ui;
+
+    public Nova(String filePath) {
+        ui = new Ui();
+        this.storage = new Storage(filePath);
+        try {
+            TaskList loaded = storage.load();
+            this.tasks = (loaded != null) ? loaded : new TaskList();
+        } catch (NovaException e) {
+            ui.showLoadingError();
```
Comment (grenn24) [inline src/main/java/nova/Nova.java:34]: Can consider printing out the error message in addition to loading failed error to help with debugging

File: src/main/java/nova/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Parses through user's input to either return a Command or throw an error for unreadable input.
+     *
+     * @param line User's input.
+     * @return Command based on user's input.
+     * @throws NovaException if user's input is unreadable.
+     */
+    public static Command parse(String line) throws NovaException {
+        String[] parts = line.trim().split(" ", 2);
+        String commandWord = parts[0].toLowerCase();
```
Comment (grenn24) [inline src/main/java/nova/parser/Parser.java:68]: You might want to handle the case where the input is empty or consists only of whitespace, since parts[0] could throw an ArrayIndexOutOfBoundsException if line.trim() is empty for example.

File: src/main/java/nova/tasks/Event.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     * Returns a string representation of the Event task.
+     * The format includes the task type identifier [E], the task description,
+     * and the formatted start and end date/time in the pattern "MMM d yyyy HHmm".
+     *
+     * @return a formatted string representation of the event task
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: "
+                + from.format(DateTimeFormatter.ofPattern("MMM d yyyy HHmm")) + " to: "
```
Comment (grenn24) [inline src/main/java/nova/tasks/Event.java:60]: Can store the date format (i.e. MMM d yyyy HHmm) as a class field constant so that its easier to reference later

File: src/main/java/nova/tasks/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * Returns a string representation of the Deadline task.
+     * The format includes the task type identifier [D], the task description,
+     * and the formatted due date/time in the pattern "MMM d yyyy HHmm".
+     *
+     * @return a formatted string representation of the deadline task
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by.format(DateTimeFormatter.ofPattern("MMM d yyyy HHmm")) + ")";
```
Comment (grenn24) [inline src/main/java/nova/tasks/Deadline.java:47]: Can store the date format (i.e. MMM d yyyy HHmm) as a class field constant so that its easier to reference later

File: src/test/java/nova/tasks/EventTest.java
Comment (grenn24) [inline src/test/java/nova/tasks/EventTest.java:1]: Can consider testing for more event edge cases, like events created with from dates after to dates.

File: src/test/java/nova/tasks/EventTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    void testToStringWhenNotDone() {
+        String expected = "[E][ ] project meeting (from: Aug 27 2025 1400 to: Aug 27 2025 1630)";
+        assertEquals(expected, event.toString(),
+                "toString() should format event correctly with status and date range");
+    }
+
+    @Test
+    void testToStringWhenDone() {
+        event.mark();
+        String expected = "[E][X] project meeting (from: Aug 27 2025 1400 to: Aug 27 2025 1630)";
```
Comment (grenn24) [inline src/test/java/nova/tasks/EventTest.java:50]: Consider using the same date format in Event class (stored as static field) to minimise inconsistencies

File: src/main/java/nova/ui/Ui.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
+package nova.ui;
+
+import java.util.Scanner;
+
+public class Ui {
+    /**
+     * Dividing line to separate Nova's responses
+     */
+    protected String LINE = "____________________________________________________________";
```
Comment (grenn24) [inline src/main/java/nova/ui/Ui.java]: should make it private static final since its effectively a constant

**Review by grenn24 — COMMENTED** [review]
Good job on the whole. 👍
There's clear separation of logic between the classes used.
Just some changes suggested to help in code styling and handling edge cases.

File: src/main/java/nova/Nova.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import javafx.application.Application;
+import nova.commands.Command;
+import nova.exceptions.NovaException;
+import nova.parser.Parser;
+import nova.storage.Storage;
+import nova.tasks.TaskList;
+import nova.ui.Main;
+import nova.ui.Ui;
+/**
```
Comment (yonglinnnnnn) [inline src/main/java/nova/Nova.java:11]: Perhaps a new line can be added before the start of documentation for the class

File: src/main/java/nova/commands/Command.java
Comment (yonglinnnnnn) [inline src/main/java/nova/commands/Command.java:1]: Perhaps consistency in documentation can be followed to all methods

File: src/main/java/nova/parser/Parser.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    private static LocalDateTime tryParseWithFormats(String dateStr, DateTimeFormatter... formatters) {
+        for (DateTimeFormatter fmt : formatters) {
+            try {
+                return LocalDateTime.parse(dateStr.trim(), fmt);
+            } catch (DateTimeParseException ignored) {
+                try {
+                    LocalDate date = LocalDate.parse(dateStr.trim(), fmt);
+                    return date.atStartOfDay();
+                } catch (DateTimeParseException ignored2) {
+                    // keep trying
```
Comment (yonglinnnnnn) [inline src/main/java/nova/parser/Parser.java:211]: Perhaps there could be a better usage of ignored and ignored2 here?

File: src/main/java/nova/storage/Storage.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                    "Failed to create data file: " + tasksFile.getAbsolutePath(), e);
+        }
+    }
+
+    /**
+     * Loads tasks from the storage file and converts them into a TaskList.
+     *
+     * @return TaskList containing tasks loaded from the file.
+     */
+    public TaskList load() {
```
Comment (yonglinnnnnn) [inline src/main/java/nova/storage/Storage.java:48]: Perhaps there could be a better method name for this?

File: src/main/java/nova/storage/Storage.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+        return loadedTasks;
+    }
+
+    /**
+     * Writes the given TaskList to the storage file.
+     *
+     * @param tasks TaskList to write to the file.
+     */
+    public void write(TaskList tasks) {
```
Comment (yonglinnnnnn) [inline src/main/java/nova/storage/Storage.java:75]: Perhaps there could be a better method name for this?

File: src/main/java/nova/tasks/TaskList.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    /**
+     * Returns the number of tasks in the list.
+     *
+     * @return the number of tasks in this TaskList
+     */
+    public int size() {
```
Comment (yonglinnnnnn) [inline src/main/java/nova/tasks/TaskList.java:39]: Perhaps there could be a better method name for this?

**Review by yonglinnnnnn — COMMENTED** [review]
LGTM. There is just some nits that can be fixed.
